lib/opentelemetry/trace/exporter/http_client.lua

```diff
@@ -28,29 +33,52 @@ function _M.new(address, timeout, headers)
         uri = uri,
         timeout = timeout,
         headers = headers,
+        httpc = httpc,
     }
     return setmetatable(self, mt)
 end
 
-function _M.do_request(self, body)
-    local httpc = http.new()
-    httpc:set_timeout(self.timeout * 1000)
+function _M.do_request(self, body, encode_gzip)
+    self.httpc = self.httpc or http.new()
 
-    local res, err = httpc:request_uri(self.uri, {
+    encode_gzip = encode_gzip or false
+    self.httpc:set_timeout(self.timeout * 1000)
+
+    if encode_gzip then
+        -- Compress (deflate) request body
+        -- the compression should be set to Best Compression and window size
+        -- should be set to 15+16, see reference below:
+        -- https://github.com/brimworks/lua-zlib/issues/4#issuecomment-26383801
+        self.headers["Content-Encoding"] = "gzip"
+        local deflate_stream = zlib.deflate(zlib.BEST_COMPRESSION, 15+16)
+        local compressed_body, _, bytes_in, bytes_out = deflate_stream(body, "finish")
+
+        otel_global.metrics_reporter:record_value(
+            exporter_request_uncompressed_payload_size, bytes_in)
+        otel_global.metrics_reporter:record_value(
+            exporter_request_compressed_payload_size, bytes_out)
+
+        body = compressed_body
+    else
+        otel_global.metrics_reporter:record_value(
+            exporter_request_uncompressed_payload_size, string.len(body))
```

Comment:
✅ 

```lua
> string.len("a")
1
> string.len("á")
2
```

Reply:
yeah, so also relevant to [your other comment](https://github.com/yangxikun/opentelemetry-lua/pull/87#discussion_r1480334357) 😂

Reply:
I'm told this is fine, `string.len` returns the length in bytes, which is exactly what we want